proposals/0032-middleware-api.md

```diff
@@ -0,0 +1,469 @@
+<!--
+  Note: You are probably looking for `stage-1--discussion-template.md`!
+  This template is reserved for anyone championing an already-approved proposal.
+
+  Community members who would like to propose an idea or feature should begin
+  by creating a GitHub Discussion. See the repo README.md for more info.
+
+  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 
+- Reference Issues: 
+  - https://github.com/withastro/roadmap/issues/531
+- Implementation PR: https://github.com/withastro/astro/pull/6721
+
+# Summary
+
+Introduce a middleware to Astro, where you can define code that runs on every request. 
+This API should work regardless of the rendering mode (SSG or SSR) or adapter used. 
+Also introduces a simple way to share request-specific data between proposed middleware, 
+API routes, and `.astro` routes.
+
+# Example
+
+The middleware pattern is useful to read `Request` and manipulate the `Response`, other than
+set and share information across endpoints and pages.
+
+
+For example, here's an example of how a user can share some information across routes:
+```js
+export const onRequest = (context, next) => {
+    if (!context.request.url.endsWith("/")) {
+        context.locals.isIndex = false;
+    } else {
+        context.locals.isIndex = true;
+    }
+}
+```
+
+Or, set some logic to make redirects:
+```js
+const redirects = new Map([
+    ["/old-1", "/new-1"],
+    ["/old-1", "/new-1"]
+])
+
+export const onRequest = (context, next) => {
+    for (const [oldRoute, newRoute] of redirects) {
+        if (context.request.url.endsWith(oldRoute)) {
+            return context.redirect(newRoute, 302);
+        }
+    }
+}
+```
+
+# Background & Motivation
+
+Middleware has been one of the most heavily requested feature in Astro. 
+It's useful for handling common tasks like auth guards and setting cache headers. 
+
+For me, it would make handling authentication much easier.
+
+
+# Goals
+
+- Provide a way intercept requests and responses, allowing users to set cookies and headers
+- Works both in SSR and SSG mode
+- Allow users to use community-created middlewares (libraries)
+  - Make available via integrations API.
+- Provide an API for request-specific data
+- Non-Node runtimes specific APIs. ie. Cloudflare Durable Objects.
+  - Add middleware from adapter.
+
+# Non-Goals
```

Comment:
You might add another section that is `# Out of Scope`. That would be things that are out of scope for this iteration of the API but are in scope for in the future such as:

- Running middleware on the edge, outside of the app itself, in production.
- Type-safe middleware, for ex. an API that would allow the types of `locals` to be mutated each time you add a new middleware to the chain.